src/controller.rs

```diff
@@ -217,21 +233,168 @@ async fn reconcile(app_instance: Arc<AppInstance>, ctx: Arc<Context>) -> Result<
                     Action::requeue(Duration::from_secs(60))
                 }
             };
-            delete_job(&app_instance, &ctx).await?;
+            delete_job(app_instance, ctx).await?;
             action
         }
     };
 
     Ok(action)
 }
 
+async fn reconcile_cleanup(app_instance: &AppInstance, ctx: &Context) -> Result<Action> {
+    info!(
+        name = app_instance.name_any(),
+        namespace = app_instance.namespace(),
+        "Cleaning up!"
+    );
+
+    info!("Deleting the running job");
+    delete_job(app_instance, ctx).await?;
```

Comment:
what if there is no running job when this code runs?

Reply:
It's okay in that case, assuming I'm understanding what you mean.

That is because this is deleting the `Job` itself, which in turn has control over the `kubit-apply-*` `Pod` which runs our apply/render steps to create the resources.

---

If you mean "what if there is no `Job` when this runs", then that would certainly be an error state. 

The `delete_job` piggybacks straight from the [`Api<Job>.delete()`
](https://docs.rs/kube/latest/kube/struct.Api.html#method.delete) function, which can return [`Err(kube_client::Error::Api)`](https://docs.rs/kube/latest/kube/enum.Error.html#variant.Api) on 4XX/5XX, in our case it'd probably be a `404`.

To guard against that cause I suppose we could quickly check for the existence of the `Job` with `get_opt` and if we get back `None` then skip over the `.delete` call.

Reply:
I think it's possible to receive a deletion event while there is no Job resource.

Checking if the resource exists is not atomic and thus would not fix the issue completely but I think it's good enough; since a) there should be only one kubit controller running at a given time and b) only the kubit controller creates the Job resource and worst case c) if the finalizer crashes it will re-run and converge